CS 598 Practical Statistical Learning Fall 2025  
Coding Assignment 2  
Group: Manas Gandhi (manaspg2), Neeya Devan, Rahul Kasibhatla  

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from abess import LinearRegression
from sklearn.metrics import mean_squared_error

# Problem 1

In [2]:
rng = np.random.default_rng(seed=773)
beta = np.array([
		1, 0.5, 0, -0.5, -1, 1, 0.5, 2, 0, 0,
		0.1, 0.2, 2, 0, 0, 0, -2, 1, 0, 0
	], dtype=float)


In [3]:
# Problem 1
#a - Generate a data set with p = 20 (independent) covariates, 
# n = 1, 000 observations and an associated quantitative response vector generated according to the model
def generate_simulated_data(n: int = 1000, p: int = 20):
    """
	Generate X (n x p), beta (p,), and Y (n,) according to the assignment specification.

	Returns:
		X: shape (n, p)
		beta: shape (p,)
		Y: shape (n,)
	"""
	#beta vector from assignment
    beta = np.array([
		1, 0.5, 0, -0.5, -1, 1, 0.5, 2, 0, 0,
		0.1, 0.2, 2, 0, 0, 0, -2, 1, 0, 0
	], dtype=float)
	
    X = rng.standard_normal(size=(n, p))
    epsilon = rng.standard_normal(n)

    Y = X.dot(beta) + epsilon
	
    return X, Y
X, Y = generate_simulated_data()

In [4]:
#b - Randomly split your data set into a training set containing 200 observations 
#    and a test set containing 800 observations.
def split_train_test(X: np.ndarray, Y: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
	indices = np.arange(X.shape[0])
	rng.shuffle(indices)
	train_indices = indices[:200]
	test_indices = indices[200:]

	X_train = X[train_indices]
	Y_train = Y[train_indices]
	
	X_test = X[test_indices]
	Y_test = Y[test_indices]
	print("X_train shape:", X_train.shape)
	print("Y_train shape:", Y_train.shape)
	print("X_test shape:", X_test.shape)
	print("Y_test shape:", Y_test.shape)
	
	return X_train, Y_train, X_test, Y_test


X_train, Y_train, X_test, Y_test = split_train_test(X, Y)

NameError: name 'Tuple' is not defined

In [ ]:
#c - Perform best subset selection on the training set, and plot the training set
# MSE associated with the best model of each size. 
def best_subset_selection(X_train: np.ndarray, Y_train: np.ndarray, p: int) -> list[float]:
	subset_mse = []
	for k in range(1, p+1):
		model = LinearRegression(support_size=k)
		model.fit(X_train, Y_train)
		yhat = model.predict(X_train)
		mse = mean_squared_error(Y_train, yhat)
		subset_mse.append(mse)

	return subset_mse

In [5]:
p=20
training_mse = best_subset_selection(X_train, Y_train, p)
plt.figure(figsize=(8, 5))
plt.plot(range(1, p + 1), training_mse, marker='o')
plt.xticks(range(1, p + 1))
plt.xlabel("Subset Size (k)")
plt.ylabel("Training MSE")
plt.title("Best Subset Selection Training MSE")
plt.grid(True)
plt.show()

NameError: name 'best_subset_selection' is not defined

In [6]:
#d. Plot the test set MSE associated with the best model of each size
def compute_mse(X_train: np.ndarray, Y_train: np.ndarray, X_test: np.ndarray, Y_test: np.ndarray, p:int) -> list[float]:
	subset_mse = []
	for k in range(1, p+1):
		model = LinearRegression(support_size=k)
		model.fit(X_train, Y_train)
		yhat = model.predict(X_test)
		mse = mean_squared_error(Y_test, yhat)
		subset_mse.append(mse)

	return subset_mse
testing_mse = compute_mse(X_train, Y_train, X_test, Y_test, p)

NameError: name 'X_train' is not defined

In [7]:
print(testing_mse)
plt.figure(figsize=(8, 5))
plt.plot(range(1, p + 1), testing_mse, marker='o')
plt.xticks(range(1, p + 1))
plt.xlabel("Subset Size (k)")
plt.ylabel("Training MSE")
plt.title("Best Subset Selection Testing MSE")
plt.grid(True)
plt.show()

NameError: name 'testing_mse' is not defined

In [8]:
#e. For which model size does the test set MSE take on its minimum value? Comment on your results.
min_mse = min(testing_mse)
best_size = testing_mse.index(min_mse) + 1
print("The best model size is " + str(best_size))

NameError: name 'testing_mse' is not defined

f - How does the model at which the test set MSE is minimized compare to the
true model used to generate the data? Comment on the coefficient values.

The test set MSE takes on its minimum value at a subset size of 11. This makes a lot of sense, given that our beta has 12 values - this means our true model size is 12, so makes sense that the ideal subset size of 11. For the nonzero predictors, the estimated coefficients are probably pretty close to their true values.

Additionally, two of the values in beta are 0.2 and 0.1, which are very small, so it's likely one of these that got missed. This is expected, however, because signals that weak are tough to reliably detect with only 200 training observations.

In [9]:
#gCreate a plot displaying error and comment on what you observe. How does this compare to the test MSE plot from (d)?
def coefficient_error(X_train: np.ndarray, Y_train: np.ndarray, beta_true: np.ndarray, p: int) -> list[float]:
    errors = []

    for r in range(1, p + 1):
        model = LinearRegression(support_size=r)
        model.fit(X_train, Y_train)
        beta_hat = model.coef_
        error = np.sum((beta_true - beta_hat) ** 2)
        errors.append(error)

    return errors
errors = coefficient_error(X_train, Y_train, beta, p)

NameError: name 'X_train' is not defined

In [10]:
print(errors)
min_error = min(errors)
best_size = errors.index(min_error) + 1
print("Lowest error is at subset size " + str(best_size))
plt.figure(figsize=(8, 5))
plt.plot(range(1, p + 1), errors, marker="o")
plt.xticks(range(1, p + 1))
plt.xlabel("Subset size (r)")
plt.ylabel(r"$\sum_{j=1}^p (\beta_j - \hat{\beta}_j^{(r)})^2$")
plt.title("Coefficient Estimation Error vs Subset Size")
plt.grid(True)
plt.show()


NameError: name 'errors' is not defined

What we can see above that there is a sharp dropoff from subset size 1-7, and reaches a minimum once again at 11. After that, the error flattens out, because the rest of the predictors are irrelevant or 0. The test MSE follows very similar behavior, reaching a minmimum at 11. This makes sense, beacuse, based on our beta array, we have 12 true nonzero coefficients, so that's around where the model should perform the best, and our error should be the lowest. As earlier, we had two very small coefficients, so the model likely had lower prediction accuracy on those because of the small training size.

# Problem 2

In [11]:
# (a) Fit a best subset selection algorithm to the data set and report the best model of each
# model size (up to 8 variables, excluding the intercept) and their prediction errors. Make
# sure that you simplify your output to only present the essential information.

import pandas as pd 
import numpy as np
import itertools
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

test_df = pd.read_csv("mls_test.csv")
train_df = pd.read_csv("mls_train.csv")

y_train = train_df["salary"]
X_train = train_df.drop(columns=["salary"])

y_test = test_df["salary"]
X_test = test_df.drop(columns=["salary"])

results = []

for k in range(1, 9): 
    best_mse = np.inf
    best_vars = None
    
    for combo in itertools.combinations(X_train.columns, k):
        model = LinearRegression()
        model.fit(X_train[list(combo)], y_train)

        y_pred = model.predict(X_test[list(combo)])
        mse = mean_squared_error(y_test, y_pred)

        if mse < best_mse:
            best_mse = mse
            best_vars = combo

    results.append({
        "Size": k,
        "Selected variables": best_vars,
        "Predicition error (mse)": best_mse
    })


results_a = pd.DataFrame(results)
print(results_a.to_string(index=False))

 Size                                                                                 Selected variables  Predicition error (mse)
    1                                                                            (ontarget_scoring_att,)             6.521076e+11
    2                                                                     (sub_on, ontarget_scoring_att)             6.295688e+11
    3                                                              (sub_on, goals, ontarget_scoring_att)             6.243642e+11
    4                                             (height, cm, weight, kg, sub_on, ontarget_scoring_att)             6.186855e+11
    5                                      (height, cm, weight, kg, sub_on, goals, ontarget_scoring_att)             6.131527e+11
    6                          (height, cm, weight, kg, sub_on, goals, aerial_won, ontarget_scoring_att)             6.091227e+11
    7           (height, cm, weight, kg, sub_on, goals, aerial_won, ontarget_scoring_att, 

In [12]:
# (b) Using the models reported in part (a), which is the best model acording to: (i) AIC, (ii)
# BIC, (iii) Cp-Mallows, and (iv) R2a? For each criterion, report the MSE for both training and testing data.

# AIC
best_aic, best_model_aic = np.inf, None
for row in results_a.itertuples(index=False):
    vars_ = list(row[1])
    k = len(vars_)
    lm = LinearRegression().fit(X_train[vars_], y_train)
    yhat_train = lm.predict(X_train[vars_])
    rss_train = np.sum((y_train - yhat_train)**2)
    curr_aic = len(y_train) * np.log(rss_train/len(y_train)) + 2*k
    
    if curr_aic < best_aic:
        best_aic = curr_aic
        best_model_aic = (vars_,
                          mean_squared_error(y_train, yhat_train),
                          mean_squared_error(y_test, lm.predict(X_test[vars_])))

print("According to AIC-- Best Model:", best_model_aic[0], "| Train MSE:", round(best_model_aic[1],3), "| Test MSE:", round(best_model_aic[2],3))


# BIC
best_bic, best_model_bic = np.inf, None
for row in results_a.itertuples(index=False):
    vars_ = list(row[1])
    k = len(vars_)
    lm = LinearRegression().fit(X_train[vars_], y_train)
    yhat_train = lm.predict(X_train[vars_])
    rss_train = np.sum((y_train - yhat_train)**2)
    bic = len(y_train) * np.log(rss_train/len(y_train)) + k*np.log(len(y_train))
    if bic < best_bic:
        best_bic = bic
        best_model_bic = (vars_,
                          mean_squared_error(y_train, yhat_train),
                          mean_squared_error(y_test, lm.predict(X_test[vars_])))

print("According to BIC-- Best Model:", best_model_bic[0], "| Train MSE:", round(best_model_bic[1],3), "| Test MSE:", round(best_model_bic[2],3))




# Cp
full_model = LinearRegression().fit(X_train, y_train)
rss_full = np.sum((y_train - full_model.predict(X_train))**2)
sigma2_hat = rss_full / (len(y_train) - X_train.shape[1] - 1)

best_cp, best_model_cp = np.inf, None
for row in results_a.itertuples(index=False):
    vars_ = list(row[1])
    k = len(vars_)
    lm = LinearRegression().fit(X_train[vars_], y_train)
    yhat_train = lm.predict(X_train[vars_])
    rss_train = np.sum((y_train - yhat_train)**2)
    cp = rss_train/sigma2_hat + 2*k - len(y_train)

    if cp < best_cp:
        best_cp = cp
        best_model_cp = (vars_,
                         mean_squared_error(y_train, yhat_train),
                         mean_squared_error(y_test, lm.predict(X_test[vars_])))

print("According to Cp-- Best Model:", best_model_cp[0],"| Train MSE:", round(best_model_cp[1],3), "| Test MSE:", round(best_model_cp[2],3))




# R2
tss = np.sum((y_train - np.mean(y_train))**2)

best_r2, best_model_r2 = -np.inf, None
for row in results_a.itertuples(index=False):
    vars_ = list(row[1])
    k = len(vars_)
    lm = LinearRegression().fit(X_train[vars_], y_train)
    yhat_train = lm.predict(X_train[vars_])
    rss_train = np.sum((y_train - yhat_train)**2)
    adj_r2 = 1 - (rss_train/(len(y_train)-k-1)) / (tss/(len(y_train)-1))
    if adj_r2 > best_r2:
        best_r2 = adj_r2
        best_model_r2 = (vars_,
                         mean_squared_error(y_train, yhat_train),
                         mean_squared_error(y_test, lm.predict(X_test[vars_])))

print("According to Adj-R2-- Best Model:", best_model_r2[0], "| Train MSE:", round(best_model_r2[1],3), "| Test MSE:", round(best_model_r2[2],3))

According to AIC-- Best Model: ['height, cm', 'weight, kg', 'sub_on', 'goals', 'yellow_card', 'won_tackle', 'aerial_won', 'ontarget_scoring_att'] | Train MSE: 296811761763.0 | Test MSE: 616410698050.296
According to BIC-- Best Model: ['sub_on', 'ontarget_scoring_att'] | Train MSE: 309434842063.795 | Test MSE: 629568772715.585
According to Cp-- Best Model: ['height, cm', 'weight, kg', 'sub_on', 'goals', 'yellow_card', 'won_tackle', 'aerial_won', 'ontarget_scoring_att'] | Train MSE: 296811761763.0 | Test MSE: 616410698050.296
According to Adj-R2-- Best Model: ['height, cm', 'weight, kg', 'sub_on', 'goals', 'yellow_card', 'won_tackle', 'aerial_won', 'ontarget_scoring_att'] | Train MSE: 296811761763.0 | Test MSE: 616410698050.296


In [13]:
# (c) Fit a ridge regression model to predict a player’s salary. Use cross-validation to select
# the best regularization parameter λ.

from sklearn.linear_model import RidgeCV
from sklearn.metrics import mean_squared_error

lambdas = np.logspace(-3, 5, 100)  

ridge_cv = RidgeCV(alphas=lambdas, store_cv_values=True)
ridge_cv.fit(X_train, y_train)

best_lambda = ridge_cv.alpha_

yhat_train = ridge_cv.predict(X_train)
yhat_test = ridge_cv.predict(X_test)
train_mse = mean_squared_error(y_train, yhat_train)
test_mse = mean_squared_error(y_test, yhat_test)

print("Best λ:", best_lambda)
print("Ridge Regression -- Train MSE:", train_mse, "| Test MSE:", test_mse)


TypeError: _BaseRidgeCV.__init__() got an unexpected keyword argument 'store_cv_values'

In [ ]:
# (d) Fit a lasso regression model on the same data set. Identify which features are shrunk to
# zero.

from sklearn.linear_model import LassoCV
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler


lambdas = np.logspace(-3, 5, 100)  

lasso_cv = make_pipeline(
    StandardScaler(),
    LassoCV(alphas=np.logspace(-3, 5, 100), cv=5, max_iter=10000, random_state=42)
)
lasso_cv.fit(X_train, y_train)

best_lambda = lasso_cv.named_steps['lassocv'].alpha_

yhat_train = lasso_cv.predict(X_train)
yhat_test = lasso_cv.predict(X_test)
train_mse = mean_squared_error(y_train, yhat_train)
test_mse = mean_squared_error(y_test, yhat_test)
coef = lasso_cv.named_steps['lassocv'].coef_
features = X_train.columns
zero_features = [f for f, c in zip(features, coef) if c == 0]

print("Best λ:", best_lambda)
print("Lasso Regression -- Train MSE:", round(train_mse,3),"| Test MSE:", round(test_mse,3))
print("Features shrunk to zero:", zero_features)


(e) Compare the performance of the models in (b) vs. ridge vs. lasso using the MSE. Which model would you recommend for predicting MLS player salaries and why?

Based on MSE, I would recommend the subset 6-varaible model from part b with an MSe OF 6.091227 × 10^11.